src/sql/passes/subqueries.rs

```diff
@@ -0,0 +1,223 @@
+
+pub trait SubQueries {
+    fn extract_and_replace_subqueries(self) -> (SqlQuery, Vec<(String, SqlQuery)>);
+}
+
+fn extract_subqueries_from_condition(ce: ConditionExpression) -> (ConditionExpression, Vec<(String, SqlQuery)>) {
+    use nom_sql::FieldExpression;
+    use nom_sql::ConditionTree;
+    use nom_sql::Column;
+
+    let mut sqs = Vec::new();
+    match ce {
+        ComparisonOp(ref ct) => {
+            let (lfq, lqueries) = extract_subqueries_from_condition(*ct.left.clone());
+            let (rfq, rqueries) = extract_subqueries_from_condition(*ct.right.clone());
+
+            sqs.extend(lqueries);
+            sqs.extend(rqueries);
+
+            (ComparisonOp(ConditionTree {
+                left: Box::new(lfq),
+                right: Box::new(rfq),
+                operator: ct.operator.clone()
+            }),
+            sqs)
+        },
+        LogicalOp(ref ct) => {
+            let (lfq, lqueries) = extract_subqueries_from_condition(*ct.left.clone());
+            let (rfq, rqueries) = extract_subqueries_from_condition(*ct.right.clone());
+
+            sqs.extend(lqueries);
+            sqs.extend(rqueries);
+
+            (LogicalOp(ConditionTree {
+                left: Box::new(lfq),
+                right: Box::new(rfq),
+                operator: ct.operator.clone()
+            }),
+            sqs)
+        },
+        NegationOp(bce) => {
+            let (expr, queries) = extract_subqueries_from_condition(*bce);
+
+            sqs.extend(queries);
+
+            (NegationOp(Box::new(expr)), sqs)
+        },
+
+        Base(NestedSelect(ref bst)) => {
+            let sq = SqlQuery::Select(*bst.clone());
+            let qname = format!("q_{}", hash_query(&sq));
+            sqs.push((qname.clone(), sq));
+
+            let cols: Vec<Column> = bst.fields.iter().map(|fe| {
+                match *fe {
+                    FieldExpression::Col(ref c) => c.clone(),
+                    _ => unimplemented!()
+                }
+            }).collect();
+            assert_eq!(cols.len(), 1);
+
+            let col = cols.first().unwrap();
+            let new_col = Column {
+                name: col.name.clone(),
+                alias: col.alias.clone(),
+                table: Some(qname),
+                function: col.function.clone(),
+            };
+
+            (Base(ConditionBase::Field(new_col.clone())), sqs)
+        },
+        Base(cb) => (Base(cb), sqs)
+    }
+}
+
+fn hash_query(q: &SqlQuery) -> u64 {
+    use std::collections::hash_map::DefaultHasher;
+    use std::hash::{Hash, Hasher};
+
+    let mut h = DefaultHasher::new();
+    q.hash(&mut h);
+    h.finish()
+}
+
+impl SubQueries for SqlQuery {
+    fn extract_and_replace_subqueries(self) -> (SqlQuery, Vec<(String, SqlQuery)>) {
+        match self {
+            SqlQuery::Select(ref st) => {
+                if st.where_clause.is_some() {
+                    let ce = st.clone().where_clause.unwrap();
```

Comment:
Is this `clone` required? `st.where_clause.as_ref().unwrap()` might work too, assuming `extract_subqueries_from_condition` can take a `&mut ConditionExpression` rather than an owned `ConditionExpression`